File: diffmerge-3/ts/src/merge.ts
import { MergeResult, MergeOptions, Conflict } from './types';
import { splitLines, normalizeLineEndings } from './utils';

/**
 * Compute LCS (Longest Common Subsequence) and return matching indices.
 */
function computeLCS(a: string[], b: string[]): [number, number][] {
  const m = a.length;
  const n = b.length;

  const dp: number[][] = Array(m + 1)
    .fill(null)
    .map(() => Array(n + 1).fill(0));

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (a[i - 1] === b[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }

  const matches: [number, number][] = [];
  let i = m;
  let j = n;

  while (i > 0 && j > 0) {
    if (a[i - 1] === b[j - 1]) {
      matches.unshift([i - 1, j - 1]);
      i--;
      j--;
    } else if (dp[i - 1][j] > dp[i][j - 1]) {
      i--;
    } else {
      j--;
    }
  }

  return matches;
}

interface Region {
  baseStart: number;
  baseEnd: number;
  content: string[];
}

/**
 * Get changed regions from base to modified version.
 */
function getChangedRegions(baseLines: string[], modLines: string[]): Region[] {
  const lcs = computeLCS(baseLines, modLines);
  const regions: Region[] = [];

  let basePos = 0;
  let modPos = 0;

  for (const [bi, mi] of lcs) {
    if (basePos < bi || modPos < mi) {
      // There's a change here
      regions.push({
        baseStart: basePos,
        baseEnd: bi,
        content: modLines.slice(modPos, mi),
      });
    }
    basePos = bi + 1;
    modPos = mi + 1;
  }

  // Handle end
  if (basePos < baseLines.length || modPos < modLines.length) {
    regions.push({
      baseStart: basePos,
      baseEnd: baseLines.length,
      content: modLines.slice(modPos),
    });
  }

  return regions;
}

/**
 * Check if two regions overlap or touch (which would cause a conflict).
 */
function regionsConflict(r1: Region, r2: Region): boolean {
  // Regions conflict if they touch or overlap in the base
  // Two regions at the same position always conflict (even if both are insertions)
  if (r1.baseStart === r2.baseStart) {
    return true;
  }
  // Otherwise check for overlap
  return !(r1.baseEnd <= r2.baseStart || r2.baseEnd <= r1.baseStart);
}

/**
 * Three-way merge with conflict detection.
 */
export function merge3(
  base: string,
  ours: string,
  theirs: string,
  options: MergeOptions = {}
): MergeResult {
  const conflictStyle = options.conflict_style ?? 'merge';
  const oursLabel = options.ours_label ?? 'ours';
  const theirsLabel = options.theirs_label ?? 'theirs';
  const baseLabel = options.base_label ?? 'base';

  const baseLines = splitLines(normalizeLineEndings(base));
  const ourLines = splitLines(normalizeLineEndings(ours));
  const theirLines = splitLines(normalizeLineEndings(theirs));

  // Get changed regions from each side
  const ourRegions = getChangedRegions(baseLines, ourLines);
  const theirRegions = getChangedRegions(baseLines, theirLines);

  // Build merged result
  const result: string[] = [];
  const conflicts: Conflict[] = [];
  let resultLineNum = 1;

  let ourIdx = 0;
  let theirIdx = 0;
  let basePos = 0;

  while (basePos <= baseLines.length || ourIdx < ourRegions.length || theirIdx < theirRegions.length) {
    const ourRegion = ourIdx < ourRegions.length ? ourRegions[ourIdx] : null;
    const theirRegion = theirIdx < theirRegions.length ? theirRegions[theirIdx] : null;

    // Check if we have a potential conflict
    if (ourRegion && theirRegion && regionsConflict(ourRegion, theirRegion)) {
      // First, copy any unchanged base lines up to the conflict start
      const conflictStart = Math.min(ourRegion.baseStart, theirRegion.baseStart);
      while (basePos < conflictStart && basePos < baseLines.length) {
        result.push(baseLines[basePos]);
        resultLineNum++;
        basePos++;
      }

      // Find the full extent of conflicting regions
      let conflictBaseEnd = Math.max(ourRegion.baseEnd, theirRegion.baseEnd);

      // Collect all our regions that are part of this conflict
      const ourContentParts: string[] = [...ourRegion.content];
      let ourEnd = ourRegion.baseEnd;
      let nextOurIdx = ourIdx + 1;

      // Collect all their regions that are part of this conflict
      const theirContentParts: string[] = [...theirRegion.content];
      let theirEnd = theirRegion.baseEnd;
      let nextTheirIdx = theirIdx + 1;

      const baseContent = baseLines.slice(conflictStart, conflictBaseEnd);
      const ourContent = ourContentParts;
      const theirContent = theirContentParts;

      // Check if both made the same change
      if (JSON.stringify(ourContent) === JSON.stringify(theirContent)) {
        result.push(...ourContent);
        resultLineNum += ourContent.length;
      } else {
        // This is a conflict
        const conflictStartLine = resultLineNum;
        result.push(`<<<<<<< ${oursLabel}`);
        resultLineNum++;
        result.push(...ourContent);
        resultLineNum += ourContent.length;
        if (conflictStyle === 'diff3') {
          result.push(`||||||| ${baseLabel}`);
          resultLineNum++;
          result.push(...baseContent);
          resultLineNum += baseContent.length;
        }
        result.push('=======');
        resultLineNum++;
        result.push(...theirContent);
        resultLineNum += theirContent.length;
        result.push(`>>>>>>> ${theirsLabel}`);
        resultLineNum++;

        conflicts.push({
          base: baseContent.join('\n'),
          ours: ourContent.join('\n'),
          theirs: theirContent.join('\n'),
          start_line: conflictStartLine,
          end_line: resultLineNum - 1,
        });
      }

      basePos = conflictBaseEnd;
      ourIdx = nextOurIdx;
      theirIdx = nextTheirIdx;
    } else if (ourRegion && (!theirRegion || ourRegion.baseStart <= theirRegion.baseStart)) {
      // Only we have a change (or our change comes first)
      // Copy base lines up to our change
      while (basePos < ourRegion.baseStart && basePos < baseLines.length) {
        result.push(baseLines[basePos]);
        resultLineNum++;
        basePos++;
      }

      // Apply our change
      result.push(...ourRegion.content);
      resultLineNum += ourRegion.content.length;
      basePos = ourRegion.baseEnd;
      ourIdx++;
    } else if (theirRegion) {
      // Only they have a change (or their change comes first)
      // Copy base lines up to their change
      while (basePos < theirRegion.baseStart && basePos < baseLines.length) {
        result.push(baseLines[basePos]);
        resultLineNum++;
        basePos++;
      }

      // Apply their change
      result.push(...theirRegion.content);
      resultLineNum += theirRegion.content.length;
      basePos = theirRegion.baseEnd;
      theirIdx++;
    } else {
      // No more changes, copy remaining base lines
      if (basePos < baseLines.length) {
        result.push(baseLines[basePos]);
        resultLineNum++;
      }
      basePos++;
    }
  }

  // Build final content
  let content = result.length > 0 ? result.join('\n') + '\n' : '';

  // Handle empty result case
  if (baseLines.length === 0 && ourLines.length === 0 && theirLines.length === 0) {
    content = '';
  }

  return {
    content,
    has_conflicts: conflicts.length > 0,
    conflicts,
  };
}

/**
 * Check if content contains conflict markers.
 */
export function hasConflicts(content: string): boolean {
  const lines = content.split('\n');
  let hasStart = false;
  let hasMiddle = false;
  let hasEnd = false;

  for (const line of lines) {
    if (line.startsWith('<<<<<<<')) {
      hasStart = true;
    } else if (line.startsWith('=======')) {
      hasMiddle = true;
    } else if (line.startsWith('>>>>>>>')) {
      hasEnd = true;
    }
  }

  return hasStart && hasMiddle && hasEnd;
}

/**
 * Extract conflict regions from merged content.
 */
export function extractConflicts(content: string): Conflict[] {
  const lines = content.split('\n');
  const conflicts: Conflict[] = [];

  let i = 0;
  while (i < lines.length) {
    if (lines[i].startsWith('<<<<<<<')) {
      const startLine = i + 1;

      // Find ours content
      const oursLines: string[] = [];
      i++;
      while (i < lines.length && !lines[i].startsWith('|||||||') && !lines[i].startsWith('=======')) {
        oursLines.push(lines[i]);
        i++;
      }

      // Handle optional base (diff3 style)
      const baseLines: string[] = [];
      if (i < lines.length && lines[i].startsWith('|||||||')) {
        i++;
        while (i < lines.length && !lines[i].startsWith('=======')) {
          baseLines.push(lines[i]);
          i++;
        }
      }

      // Skip separator
      if (i < lines.length && lines[i].startsWith('=======')) {
        i++;
      }

      // Find theirs content
      const theirsLines: string[] = [];
      while (i < lines.length && !lines[i].startsWith('>>>>>>>')) {
        theirsLines.push(lines[i]);
        i++;
      }

      const endLine = i + 1;

      conflicts.push({
        base: baseLines.join('\n'),
        ours: oursLines.join('\n'),
        theirs: theirsLines.join('\n'),
        start_line: startLine,
        end_line: endLine,
      });

      i++;
    } else {
      i++;
    }
  }

  return conflicts;
}

/**
 * Resolve a specific conflict in the content.
 */
export function resolveConflict(
  content: string,
  conflictIndex: number,
  resolution: 'ours' | 'theirs' | 'base' | string
): string {
  const lines = content.split('\n');
  const result: string[] = [];

  let conflictNum = 0;
  let i = 0;

  while (i < lines.length) {
    if (lines[i].startsWith('<<<<<<<')) {
      if (conflictNum === conflictIndex) {
        // Find the parts of this conflict
        const oursLines: string[] = [];
        const baseLines: string[] = [];
        const theirsLines: string[] = [];

        i++;
        while (i < lines.length && !lines[i].startsWith('|||||||') && !lines[i].startsWith('=======')) {
          oursLines.push(lines[i]);
          i++;
        }

        if (i < lines.length && lines[i].startsWith('|||||||')) {
          i++;
          while (i < lines.length && !lines[i].startsWith('=======')) {
            baseLines.push(lines[i]);
            i++;
          }
        }

        if (i < lines.length && lines[i].startsWith('=======')) {
          i++;
        }

        while (i < lines.length && !lines[i].startsWith('>>>>>>>')) {
          theirsLines.push(lines[i]);
          i++;
        }

        // Skip the closing marker
        i++;

        // Apply resolution
        if (resolution === 'ours') {
          result.push(...oursLines);
        } else if (resolution === 'theirs') {
          result.push(...theirsLines);
        } else if (resolution === 'base') {
          result.push(...baseLines);
        } else {
          // Custom resolution text
          const customLines = resolution.split('\n');
          // Remove trailing empty string from split if resolution ends with newline
          if (customLines[customLines.length - 1] === '') {
            customLines.pop();
          }
          result.push(...customLines);
        }

        conflictNum++;
      } else {
        // Keep this conflict as-is
        result.push(lines[i]);
        i++;
        conflictNum++;
        while (i < lines.length && !lines[i].startsWith('>>>>>>>')) {
          result.push(lines[i]);
          i++;
        }
        if (i < lines.length) {
          result.push(lines[i]);
          i++;
        }
      }
    } else {
      result.push(lines[i]);
      i++;
    }
  }

  return result.join('\n');
}

File: diffmerge-2/go/dm/merge.go
package dm

import (
	"regexp"
	"strings"
)

// Merge3 performs a three-way merge
func Merge3(base, ours, theirs string, options MergeOptions) MergeResult {
	if options.ConflictStyle == "" {
		options.ConflictStyle = "merge"
	}
	if options.OursLabel == "" {
		options.OursLabel = "ours"
	}
	if options.TheirsLabel == "" {
		options.TheirsLabel = "theirs"
	}
	if options.BaseLabel == "" {
		options.BaseLabel = "base"
	}

	baseLines := SplitLines(base)
	ourLines := SplitLines(ours)
	theirLines := SplitLines(theirs)

	// Compute diffs from base to each side
	baseLCS := lcs(baseLines, ourLines)
	baseLCS2 := lcs(baseLines, theirLines)

	// Build change maps
	ourChanges := computeChanges(baseLines, ourLines, baseLCS)
	theirChanges := computeChanges(baseLines, theirLines, baseLCS2)

	var resultLines []string
	var conflicts []Conflict
	hasConflicts := false

	baseIdx := 0
	ourIdx := 0
	theirIdx := 0

	for baseIdx < len(baseLines) || ourIdx < len(ourLines) || theirIdx < len(theirLines) {
		// Check if both sides have changes at current base position
		ourChange := getChangeAt(ourChanges, baseIdx)
		theirChange := getChangeAt(theirChanges, baseIdx)

		// Also check for overlapping changes
		if ourChange != nil && theirChange == nil {
			// Check if our change overlaps with any of their changes
			theirChange = getOverlappingChange(theirChanges, ourChange.baseStart, ourChange.baseStart+ourChange.baseCount)
		}
		if theirChange != nil && ourChange == nil {
			// Check if their change overlaps with any of our changes
			ourChange = getOverlappingChange(ourChanges, theirChange.baseStart, theirChange.baseStart+theirChange.baseCount)
		}

		if ourChange == nil && theirChange == nil {
			// No changes - take base if available
			if baseIdx < len(baseLines) {
				resultLines = append(resultLines, baseLines[baseIdx])
				baseIdx++
				ourIdx++
				theirIdx++
			} else {
				break
			}
		} else if ourChange != nil && theirChange == nil {
			// Only our side changed
			for _, line := range ourChange.newLines {
				resultLines = append(resultLines, line)
			}
			baseIdx += ourChange.baseCount
			ourIdx += len(ourChange.newLines)
			theirIdx += ourChange.baseCount
		} else if ourChange == nil && theirChange != nil {
			// Only their side changed
			for _, line := range theirChange.newLines {
				resultLines = append(resultLines, line)
			}
			baseIdx += theirChange.baseCount
			ourIdx += theirChange.baseCount
			theirIdx += len(theirChange.newLines)
		} else {
			// Both sides changed - check if same change
			if sameChange(ourChange, theirChange) {
				// Same change - take one
				for _, line := range ourChange.newLines {
					resultLines = append(resultLines, line)
				}
				baseIdx += ourChange.baseCount
				ourIdx += len(ourChange.newLines)
				theirIdx += len(theirChange.newLines)
			} else {
				// Conflict
				hasConflicts = true

				startLine := len(resultLines) + 1

				// Add conflict markers
				resultLines = append(resultLines, "<<<<<<< "+options.OursLabel)
				for _, line := range ourChange.newLines {
					resultLines = append(resultLines, line)
				}

				if options.ConflictStyle == "diff3" {
					resultLines = append(resultLines, "||||||| "+options.BaseLabel)
					for i := baseIdx; i < baseIdx+ourChange.baseCount && i < len(baseLines); i++ {
						resultLines = append(resultLines, baseLines[i])
					}
				}

				resultLines = append(resultLines, "=======")
				for _, line := range theirChange.newLines {
					resultLines = append(resultLines, line)
				}
				resultLines = append(resultLines, ">>>>>>> "+options.TheirsLabel)

				endLine := len(resultLines)

				conflicts = append(conflicts, Conflict{
					Base:      strings.Join(getBaseLines(baseLines, baseIdx, ourChange.baseCount), "\n"),
					Ours:      strings.Join(ourChange.newLines, "\n"),
					Theirs:    strings.Join(theirChange.newLines, "\n"),
					StartLine: startLine,
					EndLine:   endLine,
				})

				baseIdx += max(ourChange.baseCount, theirChange.baseCount)
				ourIdx += len(ourChange.newLines)
				theirIdx += len(theirChange.newLines)
			}
		}
	}

	// Handle trailing content
	for ourIdx < len(ourLines) {
		resultLines = append(resultLines, ourLines[ourIdx])
		ourIdx++
	}

	content := ""
	if len(resultLines) > 0 {
		content = strings.Join(resultLines, "\n") + "\n"
	}

	return MergeResult{
		Content:      content,
		HasConflicts: hasConflicts,
		Conflicts:    conflicts,
	}
}

type change struct {
	baseStart int
	baseCount int
	newLines  []string
}

func computeChanges(baseLines, newLines []string, common []string) []change {
	var changes []change

	baseIdx := 0
	newIdx := 0
	commonIdx := 0

	for commonIdx < len(common) {
		// Find next common line in base
		baseNext := -1
		for i := baseIdx; i < len(baseLines); i++ {
			if baseLines[i] == common[commonIdx] {
				baseNext = i
				break
			}
		}

		// Find next common line in new
		newNext := -1
		for i := newIdx; i < len(newLines); i++ {
			if newLines[i] == common[commonIdx] {
				newNext = i
				break
			}
		}

		// Check if there are changes before common line
		if baseNext > baseIdx || newNext > newIdx {
			changes = append(changes, change{
				baseStart: baseIdx,
				baseCount: baseNext - baseIdx,
				newLines:  newLines[newIdx:newNext],
			})
		}

		baseIdx = baseNext + 1
		newIdx = newNext + 1
		commonIdx++
	}

	// Handle remaining lines
	if baseIdx < len(baseLines) || newIdx < len(newLines) {
		changes = append(changes, change{
			baseStart: baseIdx,
			baseCount: len(baseLines) - baseIdx,
			newLines:  newLines[newIdx:],
		})
	}

	return changes
}

func getChangeAt(changes []change, baseIdx int) *change {
	for i := range changes {
		if changes[i].baseStart == baseIdx {
			return &changes[i]
		}
		if changes[i].baseStart <= baseIdx && baseIdx < changes[i].baseStart+changes[i].baseCount {
			return &changes[i]
		}
	}
	return nil
}

// changesOverlap checks if two changes overlap in base range
func changesOverlap(a, b *change) bool {
	if a == nil || b == nil {
		return false
	}
	aEnd := a.baseStart + a.baseCount
	bEnd := b.baseStart + b.baseCount
	// Overlap if ranges intersect
	return a.baseStart < bEnd && b.baseStart < aEnd
}

// getOverlappingChange finds any change that overlaps with given base range
func getOverlappingChange(changes []change, baseStart, baseEnd int) *change {
	for i := range changes {
		cEnd := changes[i].baseStart + changes[i].baseCount
		if changes[i].baseStart < baseEnd && baseStart < cEnd {
			return &changes[i]
		}
	}
	return nil
}

func sameChange(a, b *change) bool {
	if a.baseCount != b.baseCount {
		return false
	}
	if len(a.newLines) != len(b.newLines) {
		return false
	}
	for i := range a.newLines {
		if a.newLines[i] != b.newLines[i] {
			return false
		}
	}
	return true
}

func getBaseLines(baseLines []string, start, count int) []string {
	end := start + count
	if end > len(baseLines) {
		end = len(baseLines)
	}
	if start >= len(baseLines) {
		return []string{}
	}
	return baseLines[start:end]
}

// HasConflicts checks if content contains conflict markers
func HasConflicts(content string) bool {
	return strings.Contains(content, "<<<<<<<") &&
		strings.Contains(content, "=======") &&
		strings.Contains(content, ">>>>>>>")
}

// ExtractConflicts extracts conflict regions from merged content
func ExtractConflicts(content string) []Conflict {
	var conflicts []Conflict

	lines := strings.Split(content, "\n")

	oursMarkerRe := regexp.MustCompile(`^<<<<<<<`)
	baseMarkerRe := regexp.MustCompile(`^\|\|\|\|\|\|\|`)
	separatorRe := regexp.MustCompile(`^=======`)
	theirsMarkerRe := regexp.MustCompile(`^>>>>>>>`)

	i := 0
	for i < len(lines) {
		if oursMarkerRe.MatchString(lines[i]) {
			startLine := i + 1
			var oursLines, baseLines, theirsLines []string

			i++
			// Collect ours lines
			for i < len(lines) && !baseMarkerRe.MatchString(lines[i]) && !separatorRe.MatchString(lines[i]) {
				oursLines = append(oursLines, lines[i])
				i++
			}

			// Check for base section (diff3 style)
			if i < len(lines) && baseMarkerRe.MatchString(lines[i]) {
				i++
				for i < len(lines) && !separatorRe.MatchString(lines[i]) {
					baseLines = append(baseLines, lines[i])
					i++
				}
			}

			// Skip separator
			if i < len(lines) && separatorRe.MatchString(lines[i]) {
				i++
			}

			// Collect theirs lines
			for i < len(lines) && !theirsMarkerRe.MatchString(lines[i]) {
				theirsLines = append(theirsLines, lines[i])
				i++
			}

			endLine := i + 1

			conflicts = append(conflicts, Conflict{
				Base:      strings.Join(baseLines, "\n"),
				Ours:      strings.Join(oursLines, "\n"),
				Theirs:    strings.Join(theirsLines, "\n"),
				StartLine: startLine,
				EndLine:   endLine,
			})
		}
		i++
	}

	return conflicts
}

// ResolveConflict resolves a specific conflict in the content
func ResolveConflict(content string, index int, resolution string) string {
	conflicts := ExtractConflicts(content)

	if index < 0 || index >= len(conflicts) {
		return content
	}

	lines := strings.Split(content, "\n")
	conflict := conflicts[index]

	// Determine replacement content
	var replacement string
	switch resolution {
	case "ours":
		replacement = conflict.Ours
	case "theirs":
		replacement = conflict.Theirs
	case "base":
		replacement = conflict.Base
	default:
		replacement = resolution
	}

	// Find the conflict markers in the original content
	oursMarkerRe := regexp.MustCompile(`^<<<<<<<`)
	theirsMarkerRe := regexp.MustCompile(`^>>>>>>>`)

	// Find the nth conflict
	conflictCount := 0
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if oursMarkerRe.MatchString(line) {
			if conflictCount == index {
				startIdx = i
			}
			conflictCount++
		}
		if theirsMarkerRe.MatchString(line) && startIdx != -1 && endIdx == -1 {
			endIdx = i
			break
		}
	}

	if startIdx == -1 || endIdx == -1 {
		return content
	}

	// Replace the conflict
	var result []string
	result = append(result, lines[:startIdx]...)

	if replacement != "" {
		replacementLines := strings.Split(replacement, "\n")
		result = append(result, replacementLines...)
	}

	result = append(result, lines[endIdx+1:]...)

	return strings.Join(result, "\n")
}
